sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/ParquetVectorUpdaterFactory.java

```diff
@@ -0,0 +1,979 @@
+import org.apache.spark.sql.execution.datasources.DataSourceUtils;
+import org.apache.spark.sql.execution.datasources.SchemaColumnConvertNotSupportedException;
+import org.apache.spark.sql.execution.vectorized.WritableColumnVector;
+import org.apache.spark.sql.types.DataType;
+import org.apache.spark.sql.types.DataTypes;
+import org.apache.spark.sql.types.DecimalType;
+
+import java.math.BigInteger;
+import java.time.ZoneId;
+import java.time.ZoneOffset;
+import java.util.Arrays;
+
+public class ParquetVectorUpdaterFactory {
+  private static final ZoneId UTC = ZoneOffset.UTC;
+
+  private final LogicalTypeAnnotation logicalTypeAnnotation;
+  // The timezone conversion to apply to int96 timestamps. Null if no conversion.
+  private final ZoneId convertTz;
+  private final String datetimeRebaseMode;
+  private final String int96RebaseMode;
+
+  ParquetVectorUpdaterFactory(
+      LogicalTypeAnnotation logicalTypeAnnotation,
+      ZoneId convertTz,
+      String datetimeRebaseMode,
+      String int96RebaseMode) {
+    this.logicalTypeAnnotation = logicalTypeAnnotation;
+    this.convertTz = convertTz;
+    this.datetimeRebaseMode = datetimeRebaseMode;
+    this.int96RebaseMode = int96RebaseMode;
+  }
+
+  public ParquetVectorUpdater getUpdater(ColumnDescriptor descriptor, DataType sparkType) {
+    PrimitiveType.PrimitiveTypeName typeName = descriptor.getPrimitiveType().getPrimitiveTypeName();
+
+    switch (typeName) {
+      case BOOLEAN:
+        if (sparkType == DataTypes.BooleanType) {
+          return new BooleanUpdater();
+        }
+        break;
+      case INT32:
+        if (sparkType == DataTypes.IntegerType || canReadAsIntDecimal(descriptor, sparkType)) {
+          return new IntegerUpdater();
+        } else if (sparkType == DataTypes.LongType && isUnsignedIntTypeMatched(32)) {
+          // In `ParquetToSparkSchemaConverter`, we map parquet UINT32 to our LongType.
+          // For unsigned int32, it stores as plain signed int32 in Parquet when dictionary
+          // fallbacks. We read them as long values.
+          return new UnsignedIntegerUpdater();
+        } else if (sparkType == DataTypes.ByteType) {
+          return new ByteUpdater();
+        } else if (sparkType == DataTypes.ShortType) {
+          return new ShortUpdater();
+        } else if (sparkType == DataTypes.DateType) {
+          if ("CORRECTED".equals(datetimeRebaseMode)) {
+            return new IntegerUpdater();
+          } else {
+            boolean failIfRebase = "EXCEPTION".equals(datetimeRebaseMode);
+            return new IntegerWithRebaseUpdater(failIfRebase);
+          }
+        }
+        break;
+      case INT64:
+        // This is where we implement support for the valid type conversions.
+        if (sparkType == DataTypes.LongType || canReadAsLongDecimal(descriptor, sparkType)) {
+          return new LongUpdater(DecimalType.is32BitDecimalType(sparkType));
+        } else if (isUnsignedIntTypeMatched(64)) {
```

Comment:
shall we add `sparkType == DecimalType(20, 0)`?

Reply:
Good catch. Missed this.